apis/python/tests/test_dataframe_index_columns.py

```diff
@@ -54,1354 +55,2037 @@ def arrow_table():
             },
+            {},
         ],
         [
             "tsns-py-right-none-slice",
             ["tsns"],
-            [slice(np.datetime64(946684800000000002, "s"), None)],
+            None,
+            [slice(np.datetime64(946684800000000002, "ns"), None)],
             {
                 "soma_joinid": pa.array([2, 3, 4], pa.int64()),
                 "string": pa.array(["cat", "dog", "egg"], pa.large_string()),
             },
+            {},
         ],
         [
             "tsns-py-both-none-slice",
             ["tsns"],
+            None,
             [slice(None, None)],
             "default01234",
+            {},
         ],
         [
             "tsns-numpy",
             ["tsns"],
+            None,
             [
                 np.asarray(
                     [
-                        np.datetime64(946684800000000002, "s"),
-                        np.datetime64(946684800000000003, "s"),
+                        np.datetime64(946684800000000002, "ns"),
+                        np.datetime64(946684800000000003, "ns"),
                     ]
                 )
             ],
             "default23",
+            {},
         ],
         [
             "tsns-pa-array-untyped",
             ["tsns"],
+            None,
             [pa.array([946684800000000002, 946684800000000003])],
             "default23",
+            {},
         ],
         [
             "tsns-pa-array-typed",
             ["tsns"],
-            [pa.array([946684800000000002, 946684800000000003], pa.timestamp("s"))],
+            None,
+            [pa.array([946684800000000002, 946684800000000003], pa.timestamp("ns"))],
             "default23",
+            {},
         ],
         # Index by bool -- not currently supported
         [
             "BOOL-ALL",
             ["bool"],
+            None,
             [],
-            TypeError,  # TypeError: Unsupported index type bool
+            None,
+            {"create": TypeError},
         ],
     ],
 )
-def test_types(tmp_path, arrow_table, name, index_column_names, coords, expecteds):
+def test_types(
+    tmp_path,
+    arrow_table,
+    name,
+    index_column_names,
+    domains,
+    coords,
+    expecteds,
+    errors,
+):
     uri = tmp_path.as_posix()
 
-    if expecteds == TypeError:
-        with pytest.raises(expecteds):
+    if "create" in errors:
+        with pytest.raises(errors["create"]):
             soma.DataFrame.create(
                 uri,
                 schema=arrow_table.schema,
                 index_column_names=index_column_names,
+                domains=domains,
             )
         return
 
     soma.DataFrame.create(
         uri,
         schema=arrow_table.schema,
         index_column_names=index_column_names,
+        domains=domains,
     )
 
+    if "write" in errors:
+        with pytest.raises(errors["write"]):
+            with soma.DataFrame.open(uri, "w") as sdf:
+                sdf.write(arrow_table)
+        return
```

Comment:
It might be easier to follow the tests if the good cases are tested separately from each type of bad case, so there aren’t a bunch of parameters going unused in every test case.

Reply:
@thetorpedodog what does "tested separately" mean to you? I don't want to have completely separate `test_` functions if that's what you mean -- that would result in too much duplicated code.

Reply:
I actually am suggesting having different `test_types`/`test_types_create_fail`/`test_types_write_fail`/`test_types_read_fail` functions. Given that the existing test function already duplicates the creation/writing/reading code for the good and bad cases, there’s very little additional boilerplate, and the absolute line count would be significantly lower. Each test would be focused on one individual behavior rather than one big test which tries to handle all cases.

I am generally of the opinion that a little duplication of test code is preferable to having a test that behaves significantly differently depending upon the input.

Reply:
@thetorpedodog https://github.com/single-cell-data/TileDB-SOMA/pull/1027/commits/54c7f04730a380e5c433746e5d7a98a11961a2a9